### PR TITLE
CocoaPods Extension Build Support

### LINKED ISSUE #47: Handle 'sharedApplication' is unavailable: not available on iOS (App Extension)
I'm getting an error when using this library inside of an extension:

```
'sharedApplication' is unavailable: not available on iOS (App Extension) - Use view controller based solutions where appropriate instead.
```

This is a pretty widespread and documented issue, [ref](https://miqu.me/blog/2016/11/28/app-extensions-xcode-and-cocoapods-omg/).

/cc @rclee

### PR DESCRIPTION
**Pull request by dmiluski**
## 🦷 Issue:
Customer is unable to build Maps 5.9, / Events 0.10.2 if installed via CocoaPods.
- [x] https://github.com/mapbox/mapbox-events-ios/issues/47 Unable to Build run in extension (CocoaPods)
- [x] https://github.com/mapbox/mobile-telemetry/issues/754 Internal Tracking

## Why
CocoaPods is more strict than carthage where if installed via carthage one would see a warning, if installed via CocoaPods, a developer would see an error and be unable to run. This was not caught as carthage powered example shows a warning rather than a build failure error.

In this case, MapboxMobileEvents is leveraging access to sharedApplication which is not available for NSExtensions.

There is nothing preventing the SDK/Extension from building/running.
> ld: warning: linking against a dylib which is not safe for use in application extensions: /Users/danemiluski/Library/Developer/Xcode/DerivedData/MapboxMobileEvents-gtwngrjytwknvkfqciqcsoxcqjvr/Build/Products/Debug-iphonesimulator/MapboxMobileEvents.framework/MapboxMobileEvents

To enable friendlier App Extension usage though, we can signify we support app extension API only via this setting, which requires we don't require accessing SharedApplication api.

The way to ensure this SDK plays friendly when embedded in an extension is to mark extension only APIs.

## 🥅  Goal:
Provide mechanism for developer to install/run 
Provide an extensions that can be referenced in the future to validate no errors are found on run

## 🗺️ Approach
1. For each UIApplication.sharedApplication, provide an approach that plays friendly with NSExtension or UIApplication
2. Provide example extension showing a warning free build
3. Mark MapboxMobileEvents to only use extension APIs providing breaking feedback if one is used.

## 👷‍♂️ Changes
+ Leverage MMEUIApplicationWrapper to provide variations of data sources based on whether running in an App/Extension
+ Leverage responds to selector to call if running in UIApplication
+ Tests

## 📓  Notes:

Upon Testing, I did notice that the NSUserDefault persistent keys did not transfer between extension/app.  Will follow up to see if an app group in the TestHost enables the sharing. Although it is possible to share UserDefault values between extension and app, it requires an application passing in a known App Group ID to the SDK for use. This will require further discussion on actionable approaches.

## 📱 Screenshots
Demo Today Extension including MKMapView as demonstration for the possibility of including a map/telemetry gathering in extension

| Before: Project Extensions Option | After: Project Extensions Option |
| --- | --- |
| Screen Shot 2020-06-08 at 5 01 26 PM |  Screen Shot 2020-06-11 at 11 47 46 AM |

| Before: sharedApplication Errors | After: sharedApplication Errors |
| --- | --- |
| Screen Shot 2020-06-08 at 5 11 37 PM | Screen Shot 2020-06-11 at 11 55 02 AM |

| Extension Demo GUI |
| --- |
| Screen Shot 2020-06-08 at 4 27 07 PM |

### CODE REVIEW

**Comment by alfwatt** [comment]
@dmiluski I checked out the branch and everything is building OK but when I try to run the Release target I don't see the app in the "Add Widgets" list, any suggestions on debugging?

**Comment by dmiluski** [comment]
> @dmiluski I checked out the branch and everything is building OK but when I try to run the Release target I don't see the app in the "Add Widgets" list, any suggestions on debugging?

Try running the TestHost app first, then running the extension.

**Comment by rclee** [comment]
Upon revisiting the [original ticket](https://github.com/mapbox/mapbox-events-ios/issues/47). I was looking back over this [article](https://miqu.me/blog/2016/11/28/app-extensions-xcode-and-cocoapods-omg/)...

>Since Cocoapods 1.1.0, they improved integration with App extensions, and the generated project will enable the flag `APPLICATION_EXTENSION_API_ONLY` for libraries linking against an extension target. This is correct, but then you’ll start seeing compilation errors that can be a bit puzzling. Bear in mind that the code will not compile even if you don’t use the offending API. The compiler just complains that there is code that uses unavailable API.

Which may be why some users are seeing this as a build error instead of a warning.

**Comment by dmiluski** [comment]
Verified in example repo that before this change CocoaPods installed run fails, and after succeeds.
https://github.com/mapbox/twc-collab/issues/1179#issuecomment-642293701

File: Examples/Extensions/TodayWidget/Base.lproj/MainInterface.storyboard
Comment (dmiluski) [inline Examples/Extensions/TodayWidget/Base.lproj/MainInterface.storyboard:1]: This storyboard was generated as part of the addition of an example widget target to see MapboxMobileEvents running in an extension.

File: Examples/Extensions/TodayWidget/TodayViewController.m
Comment (dmiluski) [inline Examples/Extensions/TodayWidget/TodayViewController.m:43]: This is a demonstration of how a map could be added to a widget and MME SDK started.

File: Examples/Extensions/TodayWidget/TodayWidget.entitlements
Comment (dmiluski) [inline Examples/Extensions/TodayWidget/TodayWidget.entitlements:7]: This is an App Group entitlement which is now shared between Today Widget and TestHost as a mechanism for sharing data between the two targets.

This will only function as expected if the app group id is also leveraged in the Defaults(suite:)

File: MMETestHost/ViewController.m
Comment (dmiluski) [inline MMETestHost/ViewController.m]: I experienced some difficulty debugging in the App delegate so Included this ViewController class as a place to experiment with NSUserDefaults. Although it could be removed from this PR, it may prove useful in the future to experiment with data sharing potential.

File: MapboxMobileEvents.xcodeproj/project.pbxproj
Comment (dmiluski) [inline MapboxMobileEvents.xcodeproj/project.pbxproj:1584]: The majority of this project file change have to do with the addition of a Today Widget.

This particular change though is what catches on build time if we ever use apis that don't work in extensions. Which will prevent this from happening in the future.

File: MapboxMobileEvents.xcodeproj/project.pbxproj
Comment (dmiluski) [inline MapboxMobileEvents.xcodeproj/project.pbxproj:455]: Removed MMEUINavigation.

It is not used and is accessing UIApplication.shared application unnecessarily.

File: MapboxMobileEvents/MMEUIApplicationWrapper.m
Comment (dmiluski) [inline MapboxMobileEvents/MMEUIApplicationWrapper.m:47]: This class serves as the adaptor layer between the framework and the ability to access data/methods from either sharedApplication, or extension friendly sources.

Some assumption have been made to prioritize ability to build over fully flushed extenion behavior 

eg.
- backgroundTasks do not function in extensions behaving as a no-op
- values are read from the extension's Info.plist

There are options to consider for further optimizations:
- Differing behaviors for extensions:
  - delayed launch?
  - Background URLSession? Allowing for tasks to be passed off to the main application to send
  - Observing NSExtension specific notifications in case we wanted to write out data somewhere which could be read later

But for now, prioritized ability to build over flushed out behavior expectations.

**Review by rclee — APPROVED** [review]
LGTM, Thanks for putting this together Dane!

File: MapboxMobileEvents/Categories/NSUserDefaults+MMEConfiguration.m
Comment (dmiluski) [inline MapboxMobileEvents/Categories/NSUserDefaults+MMEConfiguration.m:379]: @alfwatt , 
After potential concerns about privacy setting sharing between App/Extension, opted to disable by defaults telemetry gathering in extensions until we can further reach clarity on what should be shared, as well as responsibilities to the developer to enable sharing.

File: MMETestHost/ViewController.m
Comment (alfwatt) [inline MMETestHost/ViewController.m]: Not sure where the style guide is but generally we don't commit commented out code

File: MapboxMobileEvents/MMECommonEventData.m
Comment (alfwatt) [inline MapboxMobileEvents/MMECommonEventData.m:81]: We will eventually need an application state for 'Extension' to put into location events here, so we can track telemetry coming from extensions: https://github.com/mapbox/trace-maker/issues/97#issuecomment-638330549

File: MapboxMobileEvents/MMEUIApplicationWrapper.h
Comment (alfwatt) [inline MapboxMobileEvents/MMEUIApplicationWrapper.h:7]: We will need this to work on macOS as well, `MMApplicationWrapper` is a more future-proof name

File: Tests/MMELocationManagerTests.m
Comment (alfwatt) [inline Tests/MMELocationManagerTests.m]: Let's just delete from this branch

**Review by alfwatt — APPROVED** [review]
Some suggested changes

File: Examples/Extensions/TodayWidget/TodayViewController.h
Comment (misterkerry) [inline Examples/Extensions/TodayWidget/TodayViewController.h]: nit: linter / formatter requirements?

File: MMETestHost/ViewController.m
Comment (misterkerry) [reply MMETestHost/ViewController.m]: nice @dmiluski I agree on the experiment. You can never too be cautious with NSUserDefaults.

**Review by misterkerry — APPROVED** [review]
nice progress

File: MapboxMobileEvents/MMECommonEventData.m
Comment (dmiluski) [reply MapboxMobileEvents/MMECommonEventData.m:81]: Yes, that is the line just above.
https://github.com/mapbox/mapbox-events-ios/pull/212/files#diff-016c26dee06c6b7561b0eee370142e89R79

Because this is leveraging an existing interface
https://developer.apple.com/documentation/uikit/uiapplication/1623003-applicationstate?language=objc

`@property(nonatomic, readonly) UIApplicationState applicationState;`

Where applicationState is an integer
```objective-c
typedef enum UIApplicationState : NSInteger {
    ...
} UIApplicationState;
```

It was easier to do the inspection above with an isExtension check. Is there an approach you would prefer otherwise?

File: MapboxMobileEvents/MMEUIApplicationWrapper.h
Comment (dmiluski) [reply MapboxMobileEvents/MMEUIApplicationWrapper.h:7]: Although possible, re-naming class/protocol conventions could cause pain for integration of 1.0 Beta branch in the future leading to an overly manual process.

How would you feel about holding off on this until after that is merged. Then more ideal naming/oranizing of files folders could take place such as support for SwiftPM.

File: Tests/MMELocationManagerTests.m
Comment (dmiluski) [reply Tests/MMELocationManagerTests.m]: Will do

Comment (dmiluski) [reply Tests/MMELocationManagerTests.m]: Updated

File: Examples/Extensions/TodayWidget/TodayViewController.h
Comment (dmiluski) [reply Examples/Extensions/TodayWidget/TodayViewController.h]: Good point. I've gotten used to swift lint/formatter, I'm behind on what can be leveraged in Objective-c. Will update.

**Pull request merged**
